lib/plugins/logviewer/admin.php

```diff
@@ -153,38 +126,57 @@ protected function getFacilities()
         return $facilities;
     }
 
-    private function readEndOfLogFile($logfile, $numberOfLinesLimit)
+    /**
+     * Read the lines of the logfile and return them as an array
+     *
+     * @param string $logfilePath
+     * @return array
+     */
+    protected function getLogLines($logfilePath)
     {
         $lines = [];
-        $fp = fopen($logfile, 'r');
+        $size = filesize($logfilePath);
         
-        fseek($fp, -1, SEEK_END);
-        if(fgetc($fp) == PHP_EOL) { // if the last line is EOL ignore it
-            $pos = -2;
-        } else { // otherwise start from the last character
-            $pos = -1;
-        }
-
-        $currentLine = '';
-
-        while (count($lines) < $numberOfLinesLimit && (-1 !== fseek($fp, $pos, SEEK_END))) {
-            $char = fgetc($fp);
-            if (PHP_EOL == $char) {
-                $lines[] = $currentLine;
-                $currentLine = '';
-            } else {
-                $currentLine = $char . $currentLine;
+        if ($size <= self::MAX_READ_SIZE) {
+            $lines = file($logfilePath);
+        } else {
+            $fp = fopen($logfilePath, 'r');
+            if ($fp !== null) {
+                if (fseek($fp, -self::MAX_READ_SIZE, SEEK_END) == 0) {
+                    if (false !== ($d = fread($fp, self::MAX_READ_SIZE))) {
+                        $lines = explode("\n", $d);  // Use the same line delimiter as in
+                                                     // inc/Logger.php
+                        $d = null;  // Free up space.
+
+                        // Discard the first line as it is probably truncated:
+                        array_shift($lines);
+
+                        // Discard indented lines at the start:
+                        while (!empty($lines) && (substr($lines[0], 0, 2) === '  '))
+                            array_shift($lines);
+
+                        // Add a note (as a fake log entry) to indicate that the first
+                        // log lines in the file are not shown.
+                        // The following should probably use a localizable text?
+                        $linesSkippedMsg = 'Log file too large. Previous lines skipped!';
+                        array_unshift($lines, "******\t"."\t".'['.$linesSkippedMsg.']');
+                    }
+                }
+                fclose($fp);
             }
```

Comment:
My code does not handle the case where the `fopen()` fails very well. Some sort of fake log line telling the user about the problem might be helpful.